# The R Workspace

When you launch the R workspace, you will be greeted by what is known as a REPL.
This stands for a read-evaluate-print loop.
When you write expression in the prompt and press enter, the R interpreter reads the expression, it evaluates said expression and then it prints what said expression evaluates into.
After these steps, it loops back and prompts for new inputs.
You can think of this the same way interacting with a calculator.

While it is mainly used for writing expressions, you can also input commands such as launching the help screen using `help()`, viewing demons using `demo()`, importing packages using `source()`, update packages using `update.packages()` and more.
To exit the workspace, you can use the command `q()`

## R-Expressions

To interact with the REPL, you must write complete expressions in the prompt.

In [1]:
2 + 3

[1] 5

In the example above, when we write the expression `2 + 3` and press enter, the REPL reads the expression, and if said expression is valid, it will evaluate the expression and prints the evaluation as a reply.
This reply is the line directly after the prompt `[1] 5`, The `[1]` here tells you that the output here begins with the first value in the output.
In this particular case we only have one value as an output for the evaluation.
In some cases where there are a lot of values as an output, R will organize the reply into multiple lines:

In [2]:
1000:940

[1] 1000  999  998  997  996  995  994  993  992  991  990  989  988  987  986
[16]  985  984  983  982  981  980  979  978  977  976  975  974  973  972  971
[31]  970  969  968  967  966  965  964  963  962  961  960  959  958  957  956
[46]  955  954  953  952  951  950  949  948  947  946  945  944  943  942  941
[61]  940

In the example above, expression `1000:940` generates the numbers from 1000 to 940.
Looking at the output the second line of values start with the 16th value which happens to be 985.
The third line starts with the 31st value which happens to be 985.

As mentioned before, the REPL can only evaluate your expression if said expression is valid.
Valid in this context means, that the expression syntactically correct and complete.
If you do end up with an invalid expression, R will let you know with an informative error message:

In [3]:
4 4

ERROR: Error in parse(text = x, srcfile = src): <text>:1:3: unexpected numeric constant
1: 4 4
      ^


If you end up pressing enter with an incomplete expression, R will wait for you to complete the expression in a new line as observed in the line with the `+` symbol in the example below:

In [4]:
2 + 3 -
+ 2

[1] 3

## R Atomic Datatypes

Expressions in R can come in different datatypes.
The examples we have seen so far are numeric representations known as `Doubles` which are also known as numeric.
By default numeric expressions are interpreted as doubles unless you specify otherwise.
You can check the type of the expression using the function `class()`:

In [5]:
class(12)

[1] "numeric"

In [6]:
class(3 - 4)

[1] "numeric"

If you want to specify integers, then your numbers must end with the suffix `L`:

In [7]:
class(3L - 1L)

[1] "integer"

In [8]:
class(5L)

[1] "integer"

> As it turns out, all atomic datatypes in R are vectors. Even an expression with just one value is just a vector with one element. But for now this distinction is not really important.

R also features complex numbers using the complex datatype. Any expressions written with an imaginary term will be interpreted as complex.

In [9]:
class(3 + 2i)

[1] "complex"

In [10]:
class(0i)

[1] "complex"

Non-numeric types include logicals, which can only be either `TRUE` or `FALSE`,

In [11]:
class(TRUE)

[1] "logical"

In [12]:
class(FALSE)

[1] "logical"

In [13]:
class(TRUE | FALSE)

[1] "logical"

When writing logicals, `TRUE` or `FALSE` must be written in all-caps, otherwise R will not interpret them as logicals.

In [14]:
class(True)

ERROR: Error in eval(expr, envir, enclos): object 'True' not found


In [15]:
class(true)

ERROR: Error in eval(expr, envir, enclos): object 'true' not found


R also has datatypes to represent text data.
These are called characters.
Any text that is surrounded by pairs of double quotes `"` or pairs of single quotes `"` are interpreted as character datatypes.

In [16]:
class("e")

[1] "character"

In [17]:
class("data")

[1] "character"

In [18]:
class("3 quick brown foxes?")

[1] "character"

In [19]:
class('word')

[1] "character"

In [20]:
class('{')

[1] "character"

## R Objects

R objects allow us to conveniently hold a piece of data in memory.
Any type of data can be placed into an R object, all you have to do is to assign a valid name.
For example, we can store the number `3` into the object called `x`.
We can use the `<-` operator for this.

In [21]:
x <- 3

Note that when you do this in the REPL, R doesn't reply.
This is because `x <- 3` is not an expression, therefore it doesn't evaluate into anything.

But now that we have stored `3` into the object `x`.
Anytime we use `x` in this REPL session, it will evaluate into `3`.

In [22]:
x

[1] 3

In [23]:
x + 4

[1] 7

In [24]:
x + x

[1] 6

In [25]:
x / 2

[1] 1.5

If you want to replace the value stored in an object, you can simply reassign using the `<-`:

In [26]:
x <- 21
x

[1] 21

Note that when assigning or reassigning using `<-`, you use any valid object name.
As long as the name doesn't start with a number, doesn't use the special symbols such as `+`, `-`, `*`, `/`, `^`, `$`, etc., and is not a reserved word, then it is a valid object name.
Also, you can store any valid expression into an object name.

In [27]:
y <- 2 + 3
quotient <- 70 / 3
boolean_value <- 5 < 3
date1 = as.Date("2026-6-12")